lib/hx509/revoke.c

```diff
@@ -693,6 +667,83 @@ hx509_revoke_add_crl(hx509_context context,
     return ret;
 }
 
+/**
+ * Add a CRL file or directory to the revokation context.
+ *
+ * @param context hx509 context
+ * @param ctx hx509 revokation context
+ * @param path path to file or directory that is going to be added to the context.
+ *
+ * @return An hx509 error code, see hx509_get_error_string().
+ *
+ * @ingroup hx509_revoke
+ */
+
+HX509_LIB_FUNCTION int HX509_LIB_CALL
+hx509_revoke_add_crl(hx509_context context,
+		     hx509_revoke_ctx ctx,
+		     const char *path)
+{
+    void *data;
+    size_t i;
+    int ret;
+
+    if (strncmp(path, "FILE:", 5) != 0 && strncmp(path, "DIR:", 4) != 0) {
+	hx509_set_error_string(context, 0, HX509_UNSUPPORTED_OPERATION,
+			       "unsupported type in %s", path);
+	return HX509_UNSUPPORTED_OPERATION;
+    }
+
+
+    if (strncmp(path, "FILE:", 5) == 0) {    
+	path += 5;
+	
+	ret = _hx509_revoke_add_crl(context, ctx, path);
+	
+	return ret;
+    }
+        
+    if (strncmp(path, "DIR:", 4) == 0) {
+	path += 4;
+	
+	DIR *dirp;
+	struct dirent *dir;
+	int crls = 0;
+	
+	if ((dirp = opendir(path)) == NULL) {
+	    hx509_set_error_string(context, 0, HX509_UNSUPPORTED_OPERATION,
+			           "Could not open CRL directory %s", path);
+	    return HX509_UNSUPPORTED_OPERATION;
+	}
+	
+	do {
+	    if ((dir = readdir(dirp)) != NULL) {
+		if (strcmp(dir->d_name, ".") == 0 || strcmp(dir->d_name, "..") == 0)
+		    continue;
+		    
```

Comment:
Are there other exclusions that should be enforced?

- Are there a well-known file name extensions that should be required?
- Are there illegal characters that should be excluded?
- Should backup files left by editors be excluded?
- Should files that begin with a '.' be excluded?
- Should directories be excluded?
- Should symlinks be excluded?

Reply:
- Are there a well-known file name extensions that should be required?
  - Perhaps it would be better to just allow .pem and .der?
- Are there illegal characters that should be excluded?
- Should backup files left by editors be excluded?
- Should files that begin with a '.' be excluded?
  - I believe not other Heimdal "DIR:" configuration handle those cases by excluding but I could me wrong. Ultimately it all depends on how well we want to protect the user from misconfiguration.
- Should directories be excluded?
  - We really should skip them since it makes no sense in trying to do a `_hx509_revoke_add_crl` on such entries.
- Should symlinks be excluded?
  - I believe not other Heimdal "DIR:" configuration handle those cases by excluding but I could me wrong. Ultimately it all depends on how well we want to protect the user from misconfiguration.

Reply:
@nicowilliams do you have an opinion on the inclusion / exclusion policy to apply here?

Reply:
Yes, OpenSSL does hash directories, and opens files in there like so:

```
321             {
322                 BIO_snprintf(b->data, b->max,
323                              "%s%c%08lx.%s%d", ent->dir, c, h, postfix, k);
324             }
```

where `ent->dir` is the configured directory, `c` is the path component separator character (`/` on Unix, `\` on Windows), `h` is the hash, `postfix` is `""` if looking for a trust anchor or `"r"` if looking for a CRL, and `k` is a small integer for when multiple certificates or CRLs share the same 8 hex character hash.

But we're not _searching_ for a particular CRL here.  So taking every file that a) doesn't start with `.` and b) ends in `.pem` or `.der` should be good enough!  Then we don't need to hash/rehash the directory in order to be able to use the CRLs, but on the downside, we'd be reading CRLs we might not need, thus wasting I/O and CPU cycles.  For now, however, this will do.

Reply:
I guess we need to decide whether we want `DIR` to work like OpenSSL's, in which case we should check the directory when we need it, and we should just `open()` the relevant hash rather than `readdir()` and load all.  The OpenSSL approach is more performant, but we can always switch to it later, and implementing it now might be too much work.  Thoughts?